specs/core/1_shard-data-chains.md

```diff
@@ -299,78 +343,209 @@ def compute_crosslink_data_root(blocks: Sequence[ShardBlock]) -> Hash:
+        REWARD_COEFFICIENT_BASE * BASE_REWARD_FACTOR // integer_squareroot(total_balance) // BASE_REWARDS_PER_EPOCH
+    )
 
     # Check attestations
-    attester_committee = get_persistent_committee(beacon_state, shard, block.core.slot)
+    attester_committee = get_persistent_committee(beacon_state, state.shard, block.core.slot)
     pubkeys = []
+    attestations = 0
+
     for i, index in enumerate(attester_committee):
         if block.core.attester_bitfield[i]:
             pubkeys.append(beacon_state.validators[index].pubkey)
-    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE * 2):
-        assert block.attester_bitfield[i] is False
+            add_reward(state, beacon_state, index, base_reward)
+            attestations += 1
+
+    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE):
+        assert block.core.attester_bitfield[i] is False or block.core.attester_bitfield[i] == 0  # TODO: FIX Bitvector
+
     assert bls_verify(
         pubkey=bls_aggregate_pubkeys(pubkeys),
-        message_hash=candidate.core.parent_root,
-        signature=candidate.signatures.attestation_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(candidate.core.slot))
+        message_hash=block.core.parent_root,
+        signature=block.signatures.attestation_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(block.core.slot))
     )
 
     # Check proposer
-    proposer_index = get_shard_block_proposer_index(beacon_state, shard, candidate.core.slot)
+    proposer_index = get_shard_block_proposer_index(beacon_state, state.shard, block.core.slot)
     assert proposer_index is not None
+    add_reward(state, beacon_state, proposer_index, Gwei(attestations * base_reward // PROPOSER_REWARD_QUOTIENT))
     assert bls_verify(
         pubkey=beacon_state.validators[proposer_index].pubkey,
-        message_hash=hash_tree_root(candidate.core),
-        signature=candidate.signatures.proposer_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(candidate.core.slot)),
+        message_hash=hash_tree_root(block.core),
+        signature=block.signatures.proposer_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(block.core.slot)),
     )
 
-    return True
+    # Process and update block data fees
+    add_fee(state, beacon_state, proposer_index, Gwei(state.basefee * len(block.core.data) // SHARD_BLOCK_SIZE_LIMIT))
+    QUOTIENT = SHARD_BLOCK_SIZE_LIMIT * BASEFEE_ADJUSTMENT_FACTOR
+    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
+        state.basefee += Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
+    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
+        state.basefee -= Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
```

Comment:
Although you cap `state.basefee` to be at least `1` in the next line (`state.basefee = Gwei(max(1, min(...) )`), but at this line, it seems possible if `state.basefee` is `0.9`, and then `state.basefee -= Gwei(max(1, ...)` -> `state.basefee = 0.9 - 1 = -0.1`? It is not allowable since it's an SSZ uint64 field.

Maybe use a temporary variable here?

```python
    diff_amount = Gwei(max(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
        basefee_diff = diff_amount
    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
        basefee_diff = - diff_amount
    else:
        basefee_diff = 0

    state.basefee = Gwei(max(
        1,
        min(
            EFFECTIVE_BALANCE_INCREMENT // EPOCHS_PER_SHARD_PERIOD // SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH,
            state.basefee + basefee_diff,
        )
    ))
```

Reply:
I think people will complain about even temporary variables that could be negative :smile: 

I don't know, I'm fine with any approach.